### PR TITLE
Implement (m|mun)lockall for Linux and *nix libcs

### PR DESCRIPTION
**Pull request by blblack**

### CODE REVIEW

File: lib/std/c.zig [hunk 1]
```diff
@@ -1527,6 +1527,24 @@ pub usingnamespace switch (native_os) {
     },
 };
 
+// All the Unix-likes except darwin support the call and the flags in a
+// consistent way except Linux, which has an extra flag MCL_ONFAULT and has
+// non-standard values for the MCL constants on ppc, ppc64, and sparc.
+pub usingnamespace switch (native_os) {
```
Comment (Vexu) [inline lib/std/c.zig:1533]: https://github.com/ziglang/zig/pull/19195#discussion_r1526386713 applies here as well.

**Review by Vexu — APPROVED** [review]

**Comment by blblack** [comment]
@Vexu - Sorry, my updated commit is failing tests on windows and others now.  This is because it included a smoke test in `lib/std/os/test.zig` which was excluding non-supporting platforms via `@hasDecl`, which in turn was relying on the `usingnamespace` switch to only declare these interfaces on supporting platforms.  Since the simple way of removing `usingnamespace` declared them for all targets, that doesn't work anymore (they all have the decl, and instead we get an undefined symbol at link time for the test).

I'm not yet sure how to deal with this properly, but it seems like maybe I should bring this testing topic up more-generally back over in #19214 .  I suspect none of the commits there really addressed this, because all of the cases that were platform-conditional in those commits also lacked tests.

**Comment by mikdusan** [comment]
related: andrew just opened #19352

How about this pattern, This pattern seems to work and make the test gate relatively simple:

```zig
const native_os = @import("builtin").os.tag;

// suggest put this in lib/std/c.zig between
// `pub const fstat = ....` and `pub const stat = ...`
//  many conditionals there, and in alphabetic order.
pub const mlockall = switch (native_os) {
    .linux, .freebsd, .netbsd, .openbsd, .dragonfly, .solaris => private.mlockall,
    else => {},
};

// this struct already present at bottom of lib/std/c.zig
const private = struct {
    pub extern "c" fn mlockall(flags: c_int) c_int;
};

test {
    if (@TypeOf(mlockall) == void) return error.SkipZigTest;
    // use it
    _ = &mlockall;
}
```

**Comment by blblack** [comment]
That does seem like it should work.  I do wonder though, in light of #19352, whether `if (@TypeOf(mlockall) == void)` is really any better than `if (@hasDecl(mlockall))`.  I guess for now, it wins on at least finding a way to work without `usingnamespace`.

**Comment by blblack** [comment]
Note also this should wait to rebased+amended around whatever comes of #19354 as well.

**Comment by andrewrk** [comment]
still a draft, no update in 30+ days

**Pull request closed without merging**